<a href="https://colab.research.google.com/github/Nancy-K24/Moringa-School-Week-1/blob/main/Moringa_Data_Science_Prep_W4_Independent_Project_2022_02_Nancy_Kirogo_Python_Notebook.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Electric Car-Sharing Service Company Project

###Importing our Libraries

In [ ]:
# Importing the pandas library
# 
import pandas as pd

# Importing the numpy library
#
import numpy as np

###Loading the dataset

In [ ]:
#Loading the dataset using the CSV url.
url = "/content/Autolib_dataset (2).csv"
#
df = pd.read_csv(url)
df.head(10)



Address  Cars  Bluecar counter  ...  day  hour  minute
0                 2 Avenue de Suffren     0                0  ...    8    11      43
1           145 Rue Raymond Losserand     6                6  ...    6     7      24
2    2 Avenue John Fitzgerald Kennedy     3                3  ...    3    20      14
3              51 Rue Eugène Oudiné     3                3  ...    4     4      37
4  6 avenue de la Porte de Champerret     3                3  ...    8    17      23
5                8 Boulevard Voltaire     0                0  ...    6     7       2
6                      37 rue Leblanc     0                0  ...    8    18      20
7         17 Rue des Luaps Prolongée     3                3  ...    4    22      13
8               34 avenue Jean Moulin     1                1  ...    2    22      58
9        41 boulevard de Rochechouart     6                6  ...    4    15       2

[10 rows x 25 columns]

### Accessing Information about the Dataset

In [ ]:
#Previewing the data
#
print(df)

                                 Address  Cars  ...  hour  minute
0                    2 Avenue de Suffren     0  ...    11      43
1              145 Rue Raymond Losserand     6  ...     7      24
2       2 Avenue John Fitzgerald Kennedy     3  ...    20      14
3                 51 Rue Eugène Oudiné     3  ...     4      37
4     6 avenue de la Porte de Champerret     3  ...    17      23
...                                  ...   ...  ...   ...     ...
4995                 8 avenue Ménelotte     2  ...    11      26
4996                   37 rue de Dantzig     4  ...    16      56
4997                      142 rue du Bac     1  ...     7       1
4998         2 avenue du Val de Fontenay     2  ...    17      27
4999                 5 Rue Henry Brisson     4  ...    10       2

[5000 rows x 25 columns]


In [ ]:
#Getting more information about the data set
#
df.describe()

Cars  Bluecar counter  ...         hour       minute
count  5000.000000      5000.000000  ...  5000.000000  5000.000000
mean      2.333600         2.333600  ...    11.509200    29.270000
std       2.035274         2.035274  ...     6.893549    17.231741
min       0.000000         0.000000  ...     0.000000     0.000000
25%       0.000000         0.000000  ...     6.000000    14.000000
50%       2.000000         2.000000  ...    11.000000    29.000000
75%       4.000000         4.000000  ...    18.000000    44.000000
max       7.000000         7.000000  ...    23.000000    59.000000

[8 rows x 12 columns]

In [ ]:
#Describing the data type 
#
df.dtypes # This will help reveal if there will be need for data type conversion.

Address                object
Cars                    int64
Bluecar counter         int64
Utilib counter          int64
Utilib 1.4 counter      int64
Charge Slots            int64
Charging Status        object
City                   object
Displayed comment      object
ID                     object
Kind                   object
Geo point              object
Postal code             int64
Public name            object
Rental status          object
Scheduled at           object
Slots                   int64
Station type           object
Status                 object
Subscription status    object
year                    int64
month                   int64
day                     int64
hour                    int64
minute                  int64
dtype: object

In [ ]:
#Describing the column names
#
df.columns #This will help reveal if there are white spaces that need to be stripped.

Index(['Address', 'Cars', 'Bluecar counter', 'Utilib counter',
       'Utilib 1.4 counter', 'Charge Slots', 'Charging Status', 'City',
       'Displayed comment', 'ID', 'Kind', 'Geo point', 'Postal code',
       'Public name', 'Rental status', 'Scheduled at', 'Slots', 'Station type',
       'Status', 'Subscription status', 'year', 'month', 'day', 'hour',
       'minute'],
      dtype='object')

### Data Cleaning

In [ ]:
# Counting the number of non - missing values in the df dataframe
#
df.count() #This will help reveal the coulumns that have missing values

Address                5000
Cars                   5000
Bluecar counter        5000
Utilib counter         5000
Utilib 1.4 counter     5000
Charge Slots           5000
Charging Status        5000
City                   5000
Displayed comment       111
ID                     5000
Kind                   5000
Geo point              5000
Postal code            5000
Public name            5000
Rental status          5000
Scheduled at             47
Slots                  5000
Station type           5000
Status                 5000
Subscription status    5000
year                   5000
month                  5000
day                    5000
hour                   5000
minute                 5000
dtype: int64

In [ ]:
#Determining the number of missing values per column
num_rows = df.shape[0]
num_missing = num_rows - df.count() 
num_missing

Address                   0
Cars                      0
Bluecar counter           0
Utilib counter            0
Utilib 1.4 counter        0
Charge Slots              0
Charging Status           0
City                      0
Displayed comment      4889
ID                        0
Kind                      0
Geo point                 0
Postal code               0
Public name               0
Rental status             0
Scheduled at           4953
Slots                     0
Station type              0
Status                    0
Subscription status       0
year                      0
month                     0
day                       0
hour                      0
minute                    0
dtype: int64

In [ ]:
# # Counting columns with duplicate values in the df dataframe
#
df.duplicated()
df.duplicated(subset=['Cars', 'Bluecar counter'])


0       False
1       False
2       False
3        True
4        True
        ...  
4995     True
4996     True
4997     True
4998     True
4999     True
Length: 5000, dtype: bool

In [ ]:
# Concatenating the year, month and day columns to have data in the correct format
#
df['dateInt']=df['year'].astype(str) + df['month'].astype(str).str.zfill(2)+ df['day'].astype(str).str.zfill(2)
df['date'] = pd.to_datetime(df['dateInt'], format='%Y%m%d')


In [ ]:
# Concatenating the hour and minute columns to have data in the correct format
#
df['timeInt']=df['hour'].astype(str) + ':' + df['minute'].astype(str)
df['time']=pd.to_datetime(df['timeInt'], format='%H:%M').dt.time



In [ ]:
#Dropping columns that have missing values and duplicates
#
#Dropping columns 'cars', 'Displayed comment', and 'Scheduled at'
#columns = ['Cars','Displayed comment','Scheduled at'] # Cars is a duplicate of Bluecar counter  while Displayed comment and Scheduled at have too many missing values and are not needed when analysing the data.
#df.drop('Cars', inplace=True, axis=1)
#df.drop('Displayed comment', inplace=True, axis=1)
df.drop('Scheduled at', inplace=True, axis=1)

In [ ]:
df.head(1)

Address  Cars  Bluecar counter  ...   dateInt  timeInt      time
0  2 Avenue de Suffren     0                0  ...  20180408    11:43  11:43:00

[1 rows x 29 columns]

###Data Analysis

####Identify the most popular hour of the day for picking up a shared electric car (Bluecar) in the city of Paris over the month of April 2018.

In [ ]:
#Identifying the most popular hour of the day for picking up a shared electric car (Bluecar) in the city of Paris over the month of April 2018.
#
#df['time'].count()
df['time'].groupby(df['Bluecar counter']).max()

Bluecar counter
0    23:58:00
1    23:49:00
2    23:59:00
3    23:58:00
4    23:57:00
5    23:58:00
6    23:59:00
7    21:20:00
Name: time, dtype: object

####What is the most popular hour for returning cars?

####What station is the most popular?

In [ ]:
# Overall?
#
df['Station type'].mode()


0    station
dtype: object

In [ ]:
#At the most popular picking hour?
#
df['Station type'].groupby(df['time']).max()



time
00:00:00    station
00:01:00    station
00:02:00    station
00:03:00    station
00:04:00    station
             ...   
23:55:00    station
23:56:00    station
23:57:00    station
23:58:00    station
23:59:00    station
Name: Station type, Length: 1394, dtype: object

####What postal code is the most popular for picking up Blue cars? Does the most popular station belong to that postal code?

In [ ]:
#Overall?
#
df['Postal code'].mode()

0    75015
dtype: int64

In [ ]:
#At the most popular picking hour?
#
df['Postal code'].groupby(df['time']).max()


time
00:00:00    94000
00:01:00    94110
00:02:00    93100
00:03:00    94500
00:04:00    75009
            ...  
23:55:00    94450
23:56:00    92300
23:57:00    92600
23:58:00    92600
23:59:00    92310
Name: Postal code, Length: 1394, dtype: int64

####Does the most popular station belong to that postal code?

In [ ]:
#Overall?

In [ ]:
#At the most popular picking hour?

####Do the results change if you consider Utilib and Utilib 1.4 instead of Blue cars? 